core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1311,6 +1311,51 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     */
   def mask: Stream[F, O] = this.handleErrorWith(_ => Stream.empty)
 
+  /**
+    * Like [[Stream.flatMap]] but interrupts the inner stream when new elements arrive in the outer stream.
+    *
+    * The implementation will try to preserve chunks like [[Stream.merge]].
+    *
+    * Finializers of each inner stream are guaranteed to run before the next inner stream starts.
+    *
+    * When the outer stream stops gracefully, the currently running inner stream will continue to run.
+    *
+    * When an inner stream terminates/interrupts, nothing happens until the next element arrives
+    * in the outer stream(i.e the outer stream holds the stream open during this time or else the
+    * stream terminates)
+    *
+    * When either the inner or outer stream fails, the entire stream fails and the finalizer of the
+	* inner stream runs before the outer one.
+    *
+    */
+  def switchMap[F2[x] >: F[x], O2](f: O => Stream[F2, O2])(
+      implicit F2: Concurrent[F2]): Stream[F2, O2] =
+    Stream.eval(Semaphore[F2](1)).flatMap { guard =>
+      Stream.eval(Ref.of[F2, Option[Deferred[F2, Unit]]](None)).flatMap { haltRef =>
+        Stream.eval(SignallingRef.apply(false)).flatMap { innerFailed =>
+          def runInner(o: O, halt: Deferred[F2, Unit]) =
+            Stream.eval(guard.acquire) >> // guard inner to prevent parallel inner streams
+              f(o)
+                .interruptWhen(halt.get.attempt)
+                .onError { case e => Stream.eval(innerFailed.set(true)) }
+                // upon failure, do not release the guard(prevents the next stream from running)
+                .onFinalize(innerFailed.get.flatMap { if (_) F2.unit else guard.release })
+
+          this
+            .evalMap { o =>
+              Deferred[F2, Unit].flatMap { halt =>
+                haltRef.modify {
+                  case None       => halt.some -> F2.unit
+                  case Some(last) => halt.some -> last.complete(()) // interrupt the previous one
+                }.flatten *> F2.pure(runInner(o, halt))
```

Comment:
`.flatten *> F2.pure(runInner(o, halt))` should be equivalent

`.flatten.as(runInner(o, halt))`

Reply:
Done, combined with the `getAndSet` suggested by @pchlupacek